### ISSUE TITLE
Functor and Monad

### ISSUE THREAD

**Issue by wangjiakng** [nus-cs2030/2324-s2#804]
Hii I am very confused about what is Functor and Monad and how they can be useful in coding. I will appreciate any information related to these 2 concepts! Thank you!

**Comment by jaydenntann77** [comment]
A Functor in programming is an abstraction that represents a type that can be mapped over. You can think of a Functor as any type that has a `map `method allowing you to apply a function to the contents of the container (like `Optional`, `Stream`, etc.), transforming the contents while maintaining the structure of the container.

For example:
```
Optional<String> maybeText = Optional.of("hello");

// Apply a function to the content of the Optional
Function<String, String> shout = s -> s.toUpperCase() + "!";
Optional<String> shoutedText = maybeText.map(shout);

shoutedText.ifPresent(System.out::println); // Outputs: HELLO!
```
`Optional<String>` is a container that might hold a String or be empty.
Using `map`, you apply the function shout to the content of maybeText.
The structure of `Optional `remains, but the content changes according to the function.

On the other hand, a Monad is an abstraction that deals with computation instead of values directly. It's a type of functor but with extra capabilities, typically having a `flatMap `method. This method allows for chaining operations where the result of one operation might affect the execution and type of the next operation. This is very useful for dealing with sequences of operations that include conditionality or complexity such as error handling and asynchronous computations.

**Comment by lifrocszh** [comment]
you can refer to the 3 monad laws and 2 functor laws

3 monad laws:
Left identity law - `Monad.of(x).flatMap(x -> f(x))` === `f(x)`
Right identity law - `m.flatMap(x -> Monad.of(x))` === `m`
Associative Law - `m.flatMap(x  -> f(x)).flatMap(y -> g(y))` === `m.flatMap(x -> f(x).flatMap(y -> g(y)))`

2 functor laws:
Identity law - `funct.map(x -> x)` === `funct`
Composition law - `funct.map(x -> f(g(x)))` === `funct.map(g).map(f)`

to put in general:
- functors need to have map
- monads are subset of functors, need to have flatMap and the unit method

hope this helps!

**Comment by frdyli** [comment]
**Functor:**
A Functor is a type that implements a map function, which allows you to transform the values inside the Functor while preserving its structure. In simpler terms, a Functor is any data type that you can map a function over.

Functor wraps a value (or values) inside a context (such as a container) and provides a way to apply a function to the wrapped value(s) without altering the context. 

`Optional<Integer> optionalValue = Optional.of(5);
Optional<Integer> mappedOptional = optionalValue.map(x -> x * 2);`

In the above example, map is applied to the Optional Functor, transforming the wrapped value inside the Optional without changing the context (the fact that it's an Optional).

**Monads:**
A Monad is a Functor that _also provides a way to chain computations together while handling potential failure or side effects_. Monads ensure that each computation in the chain is executed in the Monad's context (for example, Optional.flatMap(x -> x) operation ensures that the computation inside the lambda function is executed within the context of the Optional Monad)

Monads extend the concept of Functors by adding sequencing and chaining capabilities. They ensure that computations happen in a specific order and handle any side effects or failures transparently (for example by capturing the exception).
Example: Optionals, Futures/Promises (what we learn like CompletableFuture) , and Streams in Java can all be considered Monads.

`Optional<Integer> optionalValue = Optional.of(5);
Optional<Integer> result = optionalValue.flatMap(x -> {
    if (x > 0) {
        return Optional.of(x * 2);
    } else {
        return Optional.empty();
    }
});`
 
In the above example, flatMap is used to chain computations in the Optional Monad. It ensures that each computation happens within the context of the Optional, and it handles potential failure (by returning Optional.empty() when the condition is not met).

So overall, Functors help in writing cleaner, more expressive code by allowing you to apply functions to values in a context without worrying about the context itself.
Monads take it a step further by providing a way to sequence computations, handle failure or side effects, and maintain the context throughout the computation chain.

**Comment by sc2603** [comment]
You can think of functors as types that provide a map method to transform the values they contain. Functors are useful for abstracting over different types of data structures and operations, allowing you to write generic and reusable code that operates on containers of values (e.g., lists, optionals, futures) without needing to know the specific implementation details of each container.

A monad is a type that provides a way to sequence and compose computations that involve side effects or context, while also providing facilities for error handling and control flow. They are particularly useful for handling asynchronous computations, error handling, stateful computations, and other scenarios where sequencing and composition of operations are important.

**Comment by ZHD1987E** [comment]
finals is over you can close this issue
